### PR TITLE
Complete tests

### PR DESCRIPTION
**Pull request by MagicalGlove**

### CODE REVIEW

**Review by SMerved — CHANGES_REQUESTED** [review]
bad

File: Project/api/src/tests/unitTests/db/order.test/acceptCompleteAD.test.ts [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+        );
+
+        jest.restoreAllMocks();
+    });
+
+    it('complete order fail because wrong order status', async () => {
+        const orderRepository = AppDataSource.getMongoRepository(Order);
+        if (!dummyOrder?._id) throw new Error('Order was not created!');
+
+        jest.spyOn(orderRepository, 'findOne').mockResolvedValue({
```
Comment (SMerved) [inline Project/api/src/tests/unitTests/db/order.test/acceptCompleteAD.test.ts:227]: Should move mock object into a separate file, quite difficult to read.

File: Project/api/src/tests/unitTests/db/order.test/acceptReject.test.ts [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        await AppDataSource.destroy();
+    });
+
+    it('should succesfully reject order as restaurant', async () => {
+        if (!dummyOrder?._id) throw new Error('Order was not created!');
+
+        const acceptedOrder = await orderAndFeedbackRepository.acceptRejectOrder(
+            dummyOrder?._id.toString(),
+            1,
+            'Han har bestilt bearnaise pizza, ewww :vomit:'
```
Comment (SMerved) [inline Project/api/src/tests/unitTests/db/order.test/acceptReject.test.ts:58]: There is nothing wrong with bearnaise pizza, remove this heresy.

File: Project/api/src/tests/unitTests/db/order.test/calculate.test.ts
Comment (SMerved) [inline Project/api/src/tests/unitTests/db/order.test/calculate.test.ts:1]: There are a number of large mock objects in this file, should be moved into separate file for readability.
The dummyOrder reassignments could potentially be moved into a function.

File: Project/api/src/tests/unitTests/db/order.test/calculate.test.ts [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+        const feedbackData = {
+            foodRating: 5,
+            overallRating: 4,
+            deliveryRating: 3,
+            orderId: dummyOrder._id,
+        };
+
+        const feedback = await orderAndFeedbackRepository.createFeedbackAndLinkOrder(feedbackData);
+
+        dummyOrder = {
```
Comment (SMerved) [inline Project/api/src/tests/unitTests/db/order.test/calculate.test.ts:115]: Could be useful with a comment to explain the setHours data, as it is not really clear why its being set to specific values, and what it means.

File: Project/api/src/tests/unitTests/db/order.test/order.test.ts
Comment (SMerved) [inline Project/api/src/tests/unitTests/db/order.test/order.test.ts:1]: This is not part of this particular pull request, but relevant for this file:
The create order test has two large mock objects, which should be moved into a separate file for readability.
Furthermore the test creates an order from a mockOrder, but then creates a new mock order (orderData) to compare it to. It should expect the returned order from the create order function to match the initial mockOrder instead.

File: Project/api/src/tests/unitTests/db/order.test/retrieveOrders.test.ts [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+                    address: dummyOrder2.address,
+                    totalPrice: dummyOrder2.totalPrice,
+                    orderItemList: dummyOrder2.orderItemList,
+                    timestamp: dummyOrder2.timestamp,
+                }),
+            ])
+        );
+    });
+
+    it('should handle errors and return null', async () => {
```
Comment (SMerved) [inline Project/api/src/tests/unitTests/db/order.test/retrieveOrders.test.ts:101]: If the database returns an error, we shouldn't return null, we should return the error and let the caller handle it.

File: Project/api/src/tests/unitTests/db/order.test/retrieveOrders.test.ts [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+        if (!dummyOrder1.employeeID) throw new Error('Order has no employeeID!');
+
+        await expect(orderAndFeedbackRepository.GetOwnOrders('wrongID', 2)).rejects.toThrow('Error retrieving orders');
+    });
+
+    it('should get all orders basesd on restaurant id', async () => {
+        if (!dummyOrder1 || !dummyOrder2) throw new Error('An order was not created properly!');
+        if (!dummyOrder1.employeeID) throw new Error('Order has no employeeID!');
+
+        const orders = await orderAndFeedbackRepository.GetAllOrdersById('672de88ff54107237ff75565');
```
Comment (SMerved) [inline Project/api/src/tests/unitTests/db/order.test/retrieveOrders.test.ts:155]: Hardcoded id value, unclear where its coming from, and why we are certain this should exist in the database and return a successful result.

File: Project/api/src/tests/unitTests/db/restaurants.test.ts [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+import { getAllRestaurants } from '../../../RestaurantService/dbFunctions.ts';
+
+describe('Database functionality for restaurant tests', () => {
+    beforeAll(async () => {
+        await AppDataSource.initialize();
+    });
+
+    beforeEach(async () => {
+        // Declare the variables once
+        const restaurantRepository = AppDataSource.getMongoRepository(Restaurant);
```
Comment (SMerved) [inline Project/api/src/tests/unitTests/db/restaurants.test.ts:13]: Unnecesarry to initialize repositories in beforeEach, should just be done once at the top level.

File: Project/api/src/tests/unitTests/db/restaurants.test.ts
Comment (SMerved) [inline Project/api/src/tests/unitTests/db/restaurants.test.ts:1]: There is no test for if the function returns an error.

File: Project/api/src/tests/unitTests/db/user.test.ts [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+import { Address, User } from '../../../loginService/User.ts';
+
+describe('Database functionality for login tests', () => {
+    beforeAll(async () => {
+        await AppDataSource.initialize();
+    });
+
+    beforeEach(async () => {
+        // Declare the variables once
+        const userRepository = AppDataSource.getMongoRepository(User);
```
Comment (SMerved) [inline Project/api/src/tests/unitTests/db/user.test.ts:13]: Initialize at top level, not in beforeEach.

File: Project/api/src/tests/unitTests/db/user.test.ts [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+
+        // Mock Addresses
```
Comment (SMerved) [inline Project/api/src/tests/unitTests/db/user.test.ts:15]: Not an address.

File: Project/api/src/tests/unitTests/db/user.test.ts [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+
+        await userRepository.save(userCredentials);
+    });
+
+    afterAll(async () => {
+        await AppDataSource.destroy();
+    });
+
+    it('should log in and get address', async () => {
+        const addressRepository = AppDataSource.getMongoRepository(Address);
```
Comment (SMerved) [inline Project/api/src/tests/unitTests/db/user.test.ts:29]: Top level initialization

File: Project/api/src/tests/unitTests/api/order.test.ts [hunk 10]
```diff
@@ -167,8 +233,199 @@ describe('Post /acceptOrderAsDelivery', () => {
...
+        const payload = {
+            orderID: '67412feda778184dc774ecf7',
+        };
+        const response = await request(app).post('/completeOrderAsDelivery').send(payload);
+        expect(response.status).toBe(500);
+        expect(response.body).toEqual({ error: 'Error completing order: Error: Order is not at pick up stage' });
+    });
+});
+
+describe('Post /ordersById', () => {
```
[the remark this answers is not in the document]
Comment (MagicalGlove) [reply Project/api/src/tests/unitTests/api/order.test.ts]: This test exists, but was misplaced and therefore missed. It has now been moved to its appropriate placement!

File: Project/api/src/tests/unitTests/api/utilities.test.ts [hunk 11]
```diff
@@ -0,0 +1,43 @@
+import { calculateDeliveryPay } from '../../../utilities/order.ts';
+
+jest.mock('../../../RestaurantService/dbFunctions');
+
+describe('Utility functions tests', () => {
+    beforeEach(() => {
+        jest.resetAllMocks();
+    });
+
+    it('calculates the deliveree pay correctly', async () => {
```
[the remark this answers is not in the document]
Comment (MagicalGlove) [reply Project/api/src/tests/unitTests/api/utilities.test.ts:10]: The authoree disagrees 🤷‍♂️

File: Project/api/src/tests/unitTests/db/order.test/calculate.test.ts [hunk 12]
```diff
@@ -0,0 +1,341 @@
...
+                employeeID: dummyOrder.employeeID,
+                status: 3,
+            })
+        );
+    });
+
+    it('should calculate and update with moderate delivery time bonus and max TOQmultiplier', async () => {
+        const orderRepository = AppDataSource.getMongoRepository(Order);
+
+        // eslint-disable-next-line @typescript-eslint/no-unused-vars
```
Comment (SMerved) [inline Project/api/src/tests/unitTests/db/order.test/calculate.test.ts:169]: You can use _ infront of variable names to let eslint/typescript know that the variable will not be used.
Thereby removing the need for disabling eslint.

Comment (MagicalGlove) [reply Project/api/src/tests/unitTests/db/order.test/calculate.test.ts:169]: Cap

Comment (MagicalGlove) [reply Project/api/src/tests/unitTests/db/order.test/calculate.test.ts:169]: error
Cap

File: Project/api/src/tests/unitTests/db/order.test/acceptReject.test.ts (hunk 2 above)
Comment (MagicalGlove) [reply Project/api/src/tests/unitTests/db/order.test/acceptReject.test.ts:58]: 🤮

**Review by SMerved — APPROVED** [review]
good

**Pull request merged**